torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -166,45 +200,55 @@ def consolidate_state_dict(self, recipient_rank: int = 0) -> None:
         # Sync lr and other attributes in case its been updated
         self._update_param_groups()
 
-        local_cpu_state = _recursive_copy_to_device(
-            self.local_state_dict(), non_blocking=True, device=torch.device("cpu")
-        )
+        empty_messenger = torch.tensor([0], dtype=torch.uint8, device=self._device)
 
         # Pull the sharded state from all the other replicas
         # Store all the states in order, rank by rank
+
         # NOTE: In practice, `broadcast` is used, which is wasteful (gather would have been appropriate)
         # compatibility issues with some backends make the use of broadcast mandatory for now.
         # a possible follow up would be to move all sharded state management to RPC RRef
-        if self.rank == recipient_rank:
-            logging.debug("Pulling the sharded optimizer state from all replicas")
-            self._all_states = []
-            for rank in range(self.world_size):
-                if rank == self.rank:
-                    self._all_states.append(local_cpu_state)
 
-                    # Sync with other replicas, almost noop
-                    dist.broadcast_object_list([0], src=self.global_rank, group=self.group)
+        self._all_states = []
+        for rank in range(self.world_size):
+            global_rank = _get_global_rank(self.group, rank)
+
+            # This rank collects the whole state
+            if self.rank == recipient_rank:
+                if rank == self.rank:
+                    self._all_states.append(
+                        _recursive_copy_to_device(
+                            self.local_state_dict(), non_blocking=True, device=torch.device("cpu")
+                        )
+                    )
                 else:
                     # Fetch the optim state from the other replicas
-                    global_rank = _get_global_rank(self.group, rank)
-                    replica_state = [0]
-                    dist.broadcast_object_list(replica_state, src=global_rank, group=self.group)
```

Comment:
@rohan-varma I was getting deadlocks in unit tests with this configuration, with NCCL and current pytorch, I checked that all ranks were properly calling this but one of the ranks went through the call (the call returned) while the others timed out.

I checked
- that the tensors were on the gpu
- that all ranks were calling
- that the group was correct
- that the src rank was correct

Replacing this by the "homemade" (from Classy) _broadcast_object and keeping the same call parameters yields no issue. I was wondering whether you had an idea about that, something I'm doing wrong ? That worked fine with Gloo if I remember correctly

cc @mrshenli